src/Analyser/MutatingScope.php

```diff
@@ -1267,96 +1267,143 @@ private function resolveType(Expr $node): Type
 				($rightType instanceof IntegerRangeType || $rightType instanceof ConstantIntegerType || $rightType instanceof UnionType) &&
 				!($node instanceof Node\Expr\BinaryOp\Pow || $node instanceof Node\Expr\AssignOp\Pow)) {
 
-				if ($leftType instanceof ConstantIntegerType) {
-					$leftMin = $leftType->getValue();
-					$leftMax = $leftType->getValue();
-				} elseif ($leftType instanceof UnionType) {
-					$leftMin = null;
-					$leftMax = null;
-
-					foreach ($leftType->getTypes() as $type) {
-						if ($type instanceof IntegerRangeType) {
-							$leftMin = $leftMin !== null ? min($leftMin, $type->getMin()) : $type->getMin();
-							$leftMax = max($leftMax, $type->getMax());
-						} elseif ($type instanceof ConstantIntegerType) {
-							if ($node instanceof Node\Expr\BinaryOp\Minus || $node instanceof Node\Expr\AssignOp\Minus ||
-								$node instanceof Node\Expr\BinaryOp\Div || $node instanceof Node\Expr\AssignOp\Div) {
-								$leftMin = max($leftMin, $type->getValue());
-								$leftMax = $leftMax !== null ? min($leftMax, $type->getValue()) : $type->getValue();
+				/**
+				 * @param PhpParser\Node\Expr\AssignOp|PhpParser\Node\Expr\BinaryOp\BinaryOp $node
+				 * @param IntegerRangeType|ConstantIntegerType|UnionType $operand
+				 */
+				$math = static function (IntegerRangeType $range, $node, $operand) use (&$math): Type {
```

Comment:
Callback referencing itself like this leads to memory leak explosion. Please have a private method instead.